cps/main.c

```diff
@@ -16,11 +16,529 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#include <rte_kni.h>
+
+#include "gatekeeper_acl.h"
 #include "gatekeeper_cps.h"
+#include "gatekeeper_launch.h"
+#include "kni.h"
+
+/*
+ * To capture BGP packets with source port 179 or destination port 179
+ * on a global IPv6 address, we need two rules (per interface).
+ */
+#define NUM_ACL_BGP_RULES (2)
+
+#define TCP_PORT_BGP (179)
+
+/* XXX Sample parameters, need to be tested for better performance. */
+#define CPS_REQ_BURST_SIZE (32)
+
+/* Information needed to submit IPv6 BGP packets to the CPS block. */
+struct cps_bgp_req {
+	/* IPv6 BGP packets. */
+	struct rte_mbuf      *pkts[GATEKEEPER_MAX_PKT_BURST];
+
+	/* Number of packets stored in @pkts. */
+	unsigned int         num_pkts;
+
+	/* KNI that should receive @pkts. */
+	struct rte_kni       *kni;
+};
+
+/* Requests that can be made to the CPS block. */
+enum cps_req_ty {
+	/* Request to handle an IPv6 BGP packet received from another block. */
+	CPS_REQ_BGP,
+};
+
+/* Request submitted to the CPS block. */
+struct cps_request {
+	/* Type of request. */
+	enum cps_req_ty ty;
+
+	union {
+		/* If @ty is CPS_REQ_BGP, use @bgp. */
+		struct cps_bgp_req bgp;
+	} u;
+};
+
+static struct cps_config cps_conf;
+
+struct cps_config *
+get_cps_conf(void)
+{
+	return &cps_conf;
+}
+
+static int
+cleanup_cps(void)
+{
+	/* rte_kni_release() can be passed NULL. */
+	if (cps_conf.net->back_iface_enabled)
+		rte_kni_release(cps_conf.back_kni);
+	rte_kni_release(cps_conf.front_kni);
+	destroy_mailbox(&cps_conf.requests);
+	rm_kni();
+	return 0;
+}
+
+static void
+process_reqs(struct cps_config *cps_conf)
+{
+	struct cps_request *reqs[CPS_REQ_BURST_SIZE];
+	unsigned int count = mb_dequeue_burst(&cps_conf->requests,
+		(void **)reqs, CPS_REQ_BURST_SIZE);
+	unsigned int i;
+
+	for (i = 0; i < count; i++) {
+		switch (reqs[i]->ty) {
+		case CPS_REQ_BGP: {
+			struct cps_bgp_req *bgp = &reqs[i]->u.bgp;
+			unsigned int num_tx = rte_kni_tx_burst(bgp->kni,
+				bgp->pkts, bgp->num_pkts);
+			if (unlikely(num_tx < bgp->num_pkts)) {
+				uint16_t j;
+				for (j = 0; j < bgp->num_pkts - num_tx; j++)
+					rte_pktmbuf_free(bgp->pkts[j]);
+			}
+			break;
+		}
+		default:
+			RTE_LOG(ERR, GATEKEEPER,
+				"cps: unrecognized request type (%d)\n",
+				reqs[i]->ty);
+			break;
+		}
+		mb_free_entry(&cps_conf->requests, reqs[i]);
+	}
+}
+
+static void
+process_ingress(struct gatekeeper_if *iface, struct rte_kni *kni,
+	uint16_t rx_queue)
+{
+	struct rte_mbuf *bufs[GATEKEEPER_MAX_PKT_BURST];
+	uint16_t num_rx = rte_eth_rx_burst(iface->id, rx_queue, bufs,
+		GATEKEEPER_MAX_PKT_BURST);
+	unsigned int num_tx = rte_kni_tx_burst(kni, bufs, num_rx);
+
+	if (rte_kni_handle_request(kni) < 0)
+		RTE_LOG(WARNING, KNI,
+			"%s: error in handling userspace request on KNI %s\n",
+			__func__, rte_kni_get_name(kni));
+
+	if (unlikely(num_tx < num_rx)) {
+		uint16_t i;
+		for (i = 0; i < num_rx - num_tx; i++)
```

Comment:
That's the way that the DPDK sample application for the KNI does it, but I agree that it seems unnatural, and I can't find any benefit. I swapped the ordering.